From 0840c694db1dda1f0a7c5cfc3be61f40eb8f64b5 Mon Sep 17 00:00:00 2001
From: Nagamani Chennuboina <nagamani.chennuboina@intel.com>
Date: Mon, 5 Jul 2021 09:50:55 +0530
Subject: [PATCH] Changes to compile oneTBB for Android

- Added Android.bp file to create oneTBB libraries
- Changed semaphore header to point to Android bionic header

Tracked-On: OAM-97456
Signed-off-by: Nagamani Chennuboina <nagamani.chennuboina@intel.com>
---
 Android.bp          | 254 ++++++++++++++++++++++++++++++++++++++++++++
 src/tbb/semaphore.h |   4 +
 2 files changed, 258 insertions(+)
 create mode 100644 Android.bp

diff --git a/Android.bp b/Android.bp
new file mode 100644
index 00000000..26c41541
--- /dev/null
+++ b/Android.bp
@@ -0,0 +1,254 @@
+cc_library_shared {
+    name: "libtbbmalloc",
+    proprietary: true,
+    owner: "intel",
+    compile_multilib: "64",
+    defaults: ["hidl_defaults"], 
+
+    srcs: [
+        "src/tbbmalloc/backend.cpp",
+        "src/tbbmalloc/backref.cpp",
+        "src/tbbmalloc/frontend.cpp",
+        "src/tbbmalloc/large_objects.cpp",
+        "src/tbbmalloc/tbbmalloc.cpp",
+        "src/tbb/itt_notify.cpp"
+    ],
+
+    header_libs: [
+        "libtbb_headers",
+    ],
+
+    include_dirs: [
+        "bionic/libc",
+    ],
+
+    cflags: [
+        "-g",
+        "-DANDROID",
+        "-fdata-sections",
+        "-ffunction-sections",
+        "-funwind-tables",
+        "-fstack-protector-strong",
+        "-no-canonical-prefixes",
+        "-fno-addrsig",
+        "-Wa,--noexecstack",
+        "-Wformat",
+        "-Wunused-private-field",
+        "-O2",
+        "-g",
+        "-DNDEBUG",
+        "-MMD",
+        "-mrtm",
+        "-fno-rtti",
+        "-D_FORTIFY_SOURCE=2",
+        "-pthread",
+        "-std=c11",
+        "-fPIE",
+        "-fvisibility=default",
+        "-Wall",
+        "-Wno-unknown-pragmas",
+        "-Wno-strict-overflow",
+        "-Wno-unused-variable",
+        "-Wno-unused-parameter",
+        "-Wno-non-virtual-dtor",
+        "-Wno-missing-field-initializers",
+        "-Wno-error",
+        "-Wextra",
+        "-fexceptions",
+        "-fPIC"
+    ] + [
+        "-D__TBBMALLOC_BUILD",
+        "-D__TBB_USE_ITT_NOTIFY"
+    ],
+
+    shared_libs: [
+        "liblog",
+        "libc",
+        "libdl",
+        "libm"
+    ],
+
+    static_libs: []
+}
+
+cc_library_shared {
+    name: "libtbb",
+    proprietary: true,
+    owner: "intel",
+    compile_multilib: "64",
+    defaults: ["hidl_defaults"], 
+
+    srcs: [
+        "src/tbb/allocator.cpp",
+        "src/tbb/arena.cpp",
+        "src/tbb/arena_slot.cpp",
+        "src/tbb/concurrent_bounded_queue.cpp",
+        "src/tbb/dynamic_link.cpp",
+        "src/tbb/exception.cpp",
+        "src/tbb/governor.cpp",
+        "src/tbb/global_control.cpp",
+        "src/tbb/itt_notify.cpp",
+        "src/tbb/main.cpp",
+        "src/tbb/market.cpp",
+        "src/tbb/misc.cpp",
+        "src/tbb/misc_ex.cpp",
+        "src/tbb/observer_proxy.cpp",
+        "src/tbb/parallel_pipeline.cpp",
+        "src/tbb/private_server.cpp",
+        "src/tbb/profiling.cpp",
+        "src/tbb/queuing_rw_mutex.cpp",
+        "src/tbb/rml_tbb.cpp",
+        "src/tbb/rtm_mutex.cpp",
+        "src/tbb/rtm_rw_mutex.cpp",
+        "src/tbb/semaphore.cpp",
+        "src/tbb/small_object_pool.cpp",
+        "src/tbb/task.cpp",
+        "src/tbb/task_dispatcher.cpp",
+        "src/tbb/task_group_context.cpp",
+        "src/tbb/version.cpp"
+    ],
+
+    header_libs: [
+        "libtbb_headers",
+    ],
+
+    include_dirs: [
+        "bionic/libc",
+    ],
+
+    cflags: [
+        "-g",
+        "-DANDROID",
+        "-fdata-sections",
+        "-ffunction-sections",
+        "-funwind-tables",
+        "-fstack-protector-strong",
+        "-no-canonical-prefixes",
+        "-fno-addrsig",
+        "-Wa,--noexecstack",
+        "-Wformat",
+        "-Wunused-private-field",
+        "-O2",
+        "-g",
+        "-DNDEBUG",
+        "-MMD",
+        "-mrtm",
+        "-fno-rtti",
+        "-D_FORTIFY_SOURCE=2",
+        "-pthread",
+        "-std=c11",
+        "-fPIE",
+        "-fvisibility=default",
+        "-Wall",
+        "-Wno-unknown-pragmas",
+        "-Wno-strict-overflow",
+        "-Wno-unused-variable",
+        "-Wno-unused-parameter",
+        "-Wno-non-virtual-dtor",
+        "-Wno-missing-field-initializers",
+        "-Wno-error",
+        "-Wextra",
+        "-fexceptions",
+        "-fPIC",
+    ] + [
+        "-D__TBB_BUILD",
+        "-D__TBB_USE_ITT_NOTIFY"
+    ],
+
+    shared_libs: [
+        "liblog",
+        "libc",
+        "libdl",
+        "libm"
+    ],
+
+    static_libs: []
+}
+
+cc_library_shared {
+    name: "libtbbmalloc_proxy",
+    proprietary: true,
+    owner: "intel",
+    compile_multilib: "64",
+    defaults: ["hidl_defaults"], 
+
+    srcs: [
+        "src/tbbmalloc_proxy/function_replacement.cpp",
+        "src/tbbmalloc_proxy/proxy.cpp",
+    ],
+
+    header_libs: [
+        "libtbb_headers",
+    ],
+
+    include_dirs: [
+        "bionic/libc",
+    ],
+
+    cflags: [
+        "-g",
+        "-DANDROID",
+        "-fdata-sections",
+        "-ffunction-sections",
+        "-funwind-tables",
+        "-fstack-protector-strong",
+        "-no-canonical-prefixes",
+        "-fno-addrsig",
+        "-Wa,--noexecstack",
+        "-Wformat",
+        "-Wunused-private-field",
+        "-O2",
+        "-g",
+        "-DNDEBUG",
+        "-MMD",
+        "-mrtm",
+        "-fno-rtti",
+        "-D_FORTIFY_SOURCE=2",
+        "-pthread",
+        "-std=c11",
+        "-fPIE",
+        "-fvisibility=default",
+        "-Wall",
+        "-Wno-unknown-pragmas",
+        "-Wno-strict-overflow",
+        "-Wno-unused-variable",
+        "-Wno-unused-parameter",
+        "-Wno-non-virtual-dtor",
+        "-Wno-missing-field-initializers",
+        "-Wno-error",
+        "-Wextra",
+        "-fexceptions",
+        "-fPIC",
+    ] + [
+        "-D__TBBMALLOCPROXY_BUILD"
+    ],
+
+    shared_libs: [
+        "liblog",
+        "libc",
+        "libdl",
+        "libm",
+        "libtbbmalloc"
+    ],
+
+    static_libs: []
+}
+
+//#################################################
+cc_library_headers {
+    name: "libtbb_headers",
+    vendor: true,
+    export_include_dirs: [
+        "include",
+        "include/oneapi",
+        "include/oneapi/tbb/detail",
+        "include/oneapi/tbb",
+        "include/tbb",
+        "src",
+        "src/tbb",
+        "src/tbb/tools_api",
+        "src/tbb/tools_api/legacy",
+        "src/tbbmalloc",
+        "src/tbbmalloc_proxy",
+    ],
+}
diff --git a/src/tbb/semaphore.h b/src/tbb/semaphore.h
index d519b342..56b369f3 100644
--- a/src/tbb/semaphore.h
+++ b/src/tbb/semaphore.h
@@ -27,7 +27,11 @@
 #include <mach/mach_init.h>
 #include <mach/error.h>
 #else
+#ifdef __ANDROID__
+#include <include/semaphore.h>
+#else
 #include <semaphore.h>
+#endif
 #ifdef TBB_USE_DEBUG
 #include <cerrno>
 #endif
-- 
2.17.1

